### PR TITLE
CoreML: ML Program DepthToSpace

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Add DepthToSpace ML Program support.

We need to manually handle CRD format by decomposing the steps. 
Luckily we can work around CoreML's 5D limit as the data movement is not affected by merging the batch dimension with the channel dimension.

### Motivation and Context

Improve CoreML coverage for high priority models.

### CODE REVIEW

File: onnxruntime/core/providers/coreml/builders/impl/builder_utils.h [hunk 1]
```diff
@@ -136,6 +136,17 @@ void AddOperationInput(COREML_SPEC::MILSpec::Operation& op,
...
+/// <summary>
+/// Add an output to a MILSpec::Operation for an intermediate operation when the implementation is composed of
+/// multiple MLProgram operations. In this case we don't have a NodeArg for the output.
+/// </summary>
+/// <param name="op">Operation to update.</param>
+/// <param name="name">Name of the intermediate output.</param>
+/// <param name="element_type">onnx::TensorProto_DataType element type of the output.
+///   int32_t as that is what TensorShapeProto uses to store the value.</param>
+/// <param name="shape">Shape of the output if known.</param>
+void AddIntermediateOperationOutput(COREML_SPEC::MILSpec::Operation& op, const std::string& name,
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/builder_utils.h:148]: use std::string_view for consistency with other functions like AddOperationInput()?

File: onnxruntime/core/providers/coreml/builders/impl/depthtospace_op_builder.cc [hunk 2]
```diff
@@ -18,52 +19,133 @@ class DepthToSpaceOpBuilder : public BaseOpBuilder {
...
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (model_builder.CreateMLProgram()) {
+    using namespace CoreML::Specification::MILSpec;  // NOLINT
+
+    const auto mode = helper.Get("mode", "DCR");
+
+    if (mode == "DCR") {
+      // DCR is directly supported
+      // https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.tensor_transformation.depth_to_space
+      // Validated with depth_to_space.py.
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/depthtospace_op_builder.cc:45]: will this script be in another PR?

Comment (skottmckay) [reply onnxruntime/core/providers/coreml/builders/impl/depthtospace_op_builder.cc:45]: yeah - moved it to https://github.com/microsoft/onnxruntime/pull/21427 as I created a folder for the scripts

File: onnxruntime/core/providers/coreml/builders/impl/builder_utils.h (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/providers/coreml/builders/impl/builder_utils.h]: Using as the input should always be coming from ModelBuilder::GetUniqueName which returns a std::string, and we can avoid a copy when calling `set_name` this way. For AddOperationInput we're usually taking const char* input.

**Comment by skottmckay** [comment]
Included in #21472
